logsmanagement/db_api.c

```diff
@@ -0,0 +1,1096 @@
+			 * than SQLITE_DONE, which is an error then). */
+			rc = sqlite3_bind_int(stmt_retrieve_total_logs_size, 1, id);
+            if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+			rc = sqlite3_step(stmt_retrieve_total_logs_size);
+			if (rc == SQLITE_ROW){ 
+				const int64_t total_logs_filesize = (int64_t) sqlite3_column_int64(stmt_retrieve_total_logs_size, 0);
+				m_assert(total_logs_filesize == metadata_filesize, "Metadata filesize != total logs filesize");
+			}
+			else if (unlikely(rc != SQLITE_DONE)) fatal_sqlite3_err(rc, __LINE__);
+			
+			/* Get filesize of BLOB file. */ 
+			uv_fs_t stat_req;
+			rc = uv_fs_stat(db_loop, &stat_req, filename, NULL);
+			if (unlikely(rc)) fatal_libuv_err(rc, __LINE__);
+			uv_stat_t *statbuf = uv_fs_get_statbuf(&stat_req);
+			const int64_t blob_filesize = (int64_t) statbuf->st_size;
+			uv_fs_req_cleanup(&stat_req);
+			
+			do{
+				/* Case 1: blob_filesize == metadata_filesize (equal, either both zero or not): All good */
+				if(likely(blob_filesize == metadata_filesize))
+					break;
+				
+				/* Case 2: blob_filesize == 0 && metadata_filesize > 0: fatal(), however could it mean that 
+				 * EXT_BLOB_STORE_FILENAME was rotated but the SQLite metadata wasn't updated? So can it 
+				 * maybe be recovered by un-rotating? Either way, treat as fatal() for now. */
+				 // TODO: Can we avoid fatal()? 
+				if(unlikely(blob_filesize == 0 && metadata_filesize > 0)){
+					fatal("blob_filesize == 0 but metadata_filesize > 0 for '%s'\n", filename);
+				}
+				
+				/* Case 3: blob_filesize > metadata_filesize: Truncate binary to sqlite filesize, program 
+				 * crashed or terminated after writing BLOBs to external file but before metadata was updated */
+				if(unlikely(blob_filesize > metadata_filesize)){
+					infoerr("blob_filesize > metadata_filesize for '%s'. Will attempt to fix it.", filename);
+					uv_fs_t trunc_req;
+				    rc = uv_fs_ftruncate(db_loop, &trunc_req, p_file_infos_arr->data[i]->blob_handles[id], 
+				    	metadata_filesize, NULL);
+				    if(unlikely(rc)) fatal_libuv_err(rc, __LINE__);
+				    uv_fs_req_cleanup(&trunc_req);
+				    break;
+				}
+
+				/* Case 4: blob_filesize < metadata_filesize: unrecoverable (and what-should-be impossible
+				 * state), delete external binaries, clear metadata record and then fatal() */
+				// TODO: Delete external BLOB and clear metadata from DB, start from clean state but the most recent logs.
+		    	if(unlikely(blob_filesize < metadata_filesize)){
+		        	fatal("blob_filesize < metadata_filesize for '%s'. \n", filename);
+		        }
+
+			    /* Case 5: default if none of the above, should never reach here, fatal() */
+				fatal("invalid case when comparing blob_filesize with metadata_filesize");
+			} while(0);
+			
+			
+			/* Initialise blob_write_handle with logs.bin.0 */
+			if(filename[strlen(filename) - 1] == '0')
```

Comment:
Considering that `filename` was filled using `sprintf`, I suggest you to save in stack the variable length avoiding to call another function here.

Reply:
While I see your point, this code is run only once during initialisation. That's why I went with `strlen()`, since it's a case of reduced code complexity (we don't need to keep track of the variable length) over an insignificant optimisation by avoiding an `strlen()` call.